File: cpp/373.cpp
#include <iostream>
#include <queue>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution {
public:
  // copy from net
  struct hashfunction {
    size_t operator()(const pair<int, int> &x) const {
      return x.first ^ x.second;
    }
  };
  vector<vector<int>> kSmallestPairs(vector<int> &nums1, vector<int> &nums2,
                                     int k) {
    priority_queue<pair<int, pair<int, int>>> heap; // to help me sort them
    unordered_set<pair<int, int>, hashfunction> seen; // to record what I've seen
    vector<vector<int>> ret;
    int lmt = k;
    long long s1 = nums1.size(), s2 = nums2.size(); // to prevent from overflow
    if (k > s1 * s2) // let this algo run minimum round
      lmt = s1 * s2;

    // init
    heap.push(make_pair(0 - nums1[0] - nums2[0], make_pair(0, 0)));
    seen.insert(make_pair(0, 0));
    for (int i = lmt; i > 0; i--) {
      auto top = heap.top();
      ret.push_back({nums1[top.second.first], nums2[top.second.second]});
      heap.pop();

      // to add new pair in
      int i1 = top.second.first, i2 = top.second.second;
      if (i1 + 1 < s1 && seen.find({i1 + 1, i2}) == seen.end()) {
        heap.push({0 - nums1[i1 + 1] - nums2[i2], {i1 + 1, i2}});
        seen.insert({i1 + 1, i2});
      }
      if (i2 + 1 < s2 && seen.find({i1, i2 + 1}) == seen.end()) {
        heap.push({0 - nums1[i1] - nums2[i2 + 1], {i1, i2 + 1}});
        seen.insert({i1, i2 + 1});
      }
    }

    return ret;
  }
};

File: go/373.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

type Heap [][]int // structure: [[sum, idx1, idx2]]

// Len implements heap.Interface
func (h *Heap) Len() int {
	return len(*h)
}

// Less implements heap.Interface
func (h *Heap) Less(i int, j int) bool {
	return (*h)[i][0] < (*h)[j][0]
}

// Swap implements heap.Interface
func (h *Heap) Swap(i int, j int) {
	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
}

// Pop implements heap.Interface
func (h *Heap) Pop() interface{} {
	ret := (*h)[len(*h)-1]
	*h = (*h)[:len(*h)-1]
	return ret
}

// Push implements heap.Interface
func (h *Heap) Push(x interface{}) {
	*h = append(*h, x.([]int))
}

func kSmallestPairs(nums1 []int, nums2 []int, k int) [][]int {

	var h *Heap = &Heap{}
	var seen map[[2]int]int = make(map[[2]int]int, 0)
	heap.Init(h)

	var l1, l2 = len(nums1), len(nums2)
	var lmt int = int(math.Min(float64(k), float64(l1*l2)))
	heap.Push(h, []int{nums1[0] + nums2[0], 0, 0})
	seen[[2]int{0, 0}] = 1
	var ret [][]int = make([][]int, 0)
	for ; lmt > 0; lmt-- {
		top := heap.Pop(h).([]int)
		i, j := top[1], top[2]
		ret = append(ret, []int{nums1[i], nums2[j]})
		if i+1 < l1 && seen[[2]int{i + 1, j}] == 0 {
			heap.Push(h, []int{nums1[i+1] + nums2[j], i + 1, j})
			seen[[2]int{i + 1, j}]++
		}
		if j+1 < l2 && seen[[2]int{i, j + 1}] == 0 {
			heap.Push(h, []int{nums1[i] + nums2[j+1], i, j + 1})
			seen[[2]int{i, j + 1}]++
		}
	}
	return ret
}

func main() {
	fmt.Print(kSmallestPairs([]int{1, 7, 11}, []int{2, 4, 6}, 3))
}
